File: utils/liblzg/liblzg-1.0.6/src/extra/lzgmini.c
/* -*- mode: c; tab-width: 4; indent-tabs-mode: nil; -*- */

/*
* This file is part of liblzg.
*
* Copyright (c) 2010 Marcus Geelnard
*
* This software is provided 'as-is', without any express or implied
* warranty. In no event will the authors be held liable for any damages
* arising from the use of this software.
*
* Permission is granted to anyone to use this software for any purpose,
* including commercial applications, and to alter it and redistribute it
* freely, subject to the following restrictions:
*
* 1. The origin of this software must not be misrepresented; you must not
*    claim that you wrote the original software. If you use this software
*    in a product, an acknowledgment in the product documentation would
*    be appreciated but is not required.
*
* 2. Altered source versions must be plainly marked as such, and must not
*    be misrepresented as being the original software.
*
* 3. This notice may not be removed or altered from any source
*    distribution.
*/

#include <lzg.h>

/*-- PRIVATE -----------------------------------------------------------------*/

/* Internal definitions */
#define LZG_HEADER_SIZE 16
#define LZG_METHOD_COPY 0
#define LZG_METHOD_LZG1 1

/* Endian and alignment independent reader for 32-bit integers */
#define _LZG_GetUINT32(in, offs) \
    ((((lzg_uint32_t)in[offs]) << 24) | \
     (((lzg_uint32_t)in[offs+1]) << 16) | \
     (((lzg_uint32_t)in[offs+2]) << 8) | \
     ((lzg_uint32_t)in[offs+3]))

/* Calculate the checksum */
static lzg_uint32_t _LZG_CalcChecksum(const unsigned char *data, lzg_uint32_t size)
{
    unsigned short a = 1, b = 0;
    unsigned char *end = (unsigned char *)data + size;
    while (data != end)
    {
        a += *data++;
        b += a;
    }
    return (((lzg_uint32_t)b) << 16) | a;
}

/* LUT for decoding the copy length parameter */
static const unsigned char _LZG_LENGTH_DECODE_LUT[32] = {
    2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,
    18,19,20,21,22,23,24,25,26,27,28,29,35,48,72,128
};


/*-- PUBLIC ------------------------------------------------------------------*/

lzg_uint32_t LZG_DecodedSize(const unsigned char *in, lzg_uint32_t insize)
{
    /* Check header */
    if ((insize < 7) || (in[0] != 'L') || (in[1] != 'Z') || (in[2] != 'G'))
        return 0;

    /* Get output buffer size */
    return _LZG_GetUINT32(in, 3);
}

lzg_uint32_t LZG_Decode(const unsigned char *in, lzg_uint32_t insize,
    unsigned char *out, lzg_uint32_t outsize)
{
    unsigned char *src, *inEnd, *dst, *outEnd, *copy, symbol, b, b2;
    unsigned char m1, m2, m3, m4, method;
    lzg_uint32_t  i, length, offset, encodedSize, decodedSize, checksum;

    /* Check magic ID */
    if ((insize < LZG_HEADER_SIZE) || (in[0] != 'L') || (in[1] != 'Z') || (in[2] != 'G'))
        return 0;

    /* Get header data */
    decodedSize = _LZG_GetUINT32(in, 3);
    encodedSize = _LZG_GetUINT32(in, 7);
    checksum = _LZG_GetUINT32(in, 11);

    /* Check sizes */
    if ((outsize < decodedSize) || (encodedSize != (insize - LZG_HEADER_SIZE)))
        return 0;

    /* Check checksum */
    if (_LZG_CalcChecksum(&in[LZG_HEADER_SIZE], encodedSize) != checksum)
        return 0;

    /* Initialize the byte streams */
    src = (unsigned char *)in + LZG_HEADER_SIZE;;
    inEnd = ((unsigned char *)in) + insize;
    dst = out;
    outEnd = out + outsize;

    /* Check which method to use */
    method = in[15];
    if (method == LZG_METHOD_LZG1)
    {
        if (!((src + 4) <= inEnd)) return 0;
        m1 = *src++; m2 = *src++; m3 = *src++; m4 = *src++;

        /* Main decompression loop */
        while (src < inEnd)
        {
            symbol = *src++;

            if ((symbol != m1) && (symbol != m2) && (symbol != m3) && (symbol != m4))
            {
                /* Literal copy */
                if (!(dst < outEnd)) return 0;
                *dst++ = symbol;
            }
            else
            {
                /* Decode offset / length parameters */
                if (!(src < inEnd)) return 0;
                if ((b = *src++))
                {
                    if (symbol == m1)
                    {
                        /* Distant copy */
                        if (!((src + 2) <= inEnd)) return 0;
                        length = _LZG_LENGTH_DECODE_LUT[b & 0x1f];
                        b2 = *src++;
                        offset = (((unsigned int)(b & 0xe0)) << 11) |
                                  (((unsigned int)b2) << 8) |
                                  (*src++);
                        offset += 2056;
                    }
                    else if (symbol == m2)
                    {
                        /* Medium copy */
                        if (!(src < inEnd)) return 0;
                        length = _LZG_LENGTH_DECODE_LUT[b & 0x1f];
                        b2 = *src++;
                        offset = (((unsigned int)(b & 0xe0)) << 3) | b2;
                        offset += 8;
                    }
                    else if (symbol == m3)
                    {
                        /* Short copy */
                        length = (b >> 6) + 3;
                        offset = (b & 0x3f) + 8;
                    }
                    else
                    {
                        /* Near copy (including RLE) */
                        length = _LZG_LENGTH_DECODE_LUT[b & 0x1f];
                        offset = (b >> 5) + 1;
                    }

                    /* Copy the corresponding data from the history window */
                    copy = dst - offset;
                    if (!((copy >= out) && ((dst + length) <= outEnd))) return 0;
                    for (i = 0; i < length; ++i)
                        *dst++ = *copy++;
                }
                else
                {
                    /* Literal copy (single occurance of a marker symbol) */
                    if (!(dst < outEnd)) return 0;
                    *dst++ = symbol;
                }
            }
        }
    }
    else if (method == LZG_METHOD_COPY)
    {
        /* Plain copy */
        while ((src < inEnd) && (dst < outEnd))
            *dst++ = *src++;
    }

    /* All OK? */
    if ((unsigned int)(dst - out) != decodedSize)
        return 0;
    else
        return decodedSize;
}

File: utils/liblzg/liblzg-1.0.6/src/extra/lzgmini.pas
(* -*- mode: pascal; tab-width: 2; indent-tabs-mode: nil; -*- *)
unit lzgmini;

(*
* This file is part of liblzg.
*
* Copyright (c) 2010 Marcus Geelnard
*
* This software is provided 'as-is', without any express or implied
* warranty. In no event will the authors be held liable for any damages
* arising from the use of this software.
*
* Permission is granted to anyone to use this software for any purpose,
* including commercial applications, and to alter it and redistribute it
* freely, subject to the following restrictions:
*
* 1. The origin of this software must not be misrepresented; you must not
*    claim that you wrote the original software. If you use this software
*    in a product, an acknowledgment in the product documentation would
*    be appreciated but is not required.
*
* 2. Altered source versions must be plainly marked as such, and must not
*    be misrepresented as being the original software.
*
* 3. This notice may not be removed or altered from any source
*    distribution.
*)

(*
* Description
* -----------
* This is a Pascal implemention of the LZG decoder. It has been tested with the
* Free Pascal compiler (http://www.freepascal.org), and should work with
* Delphi as well (possibly with some modifications?).
*
* This version of the decoder is not quite as optimized as the routine in the
* original LZG library, but may be preferred nevertheless in Pascal
* environments.
*)


interface

type
  PByte = ^Byte;

function LZG_DecodedSize(const data: PByte; size: Cardinal): Cardinal;
function LZG_Decode(const indata: PByte; insize: Cardinal; outdata: PByte; outsize: Cardinal): Cardinal;


implementation

const
  { Internal definitions }
  LZG_HEADER_SIZE = 16;
  LZG_METHOD_COPY = 0;
  LZG_METHOD_LZG1 = 1;

  { LUT for decoding the copy length parameter }
  LZG_LENGTH_DECODE_LUT: array[0..31] of Byte = (
    2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,
    18,19,20,21,22,23,24,25,26,27,28,29,35,48,72,128
  );


{ Endian and alignment independent reader for 32-bit integers }
function LZG_GetUINT32(data: PByte; offset: Cardinal): Cardinal;
begin
  LZG_GetUINT32 := (Cardinal(data[offset]) << 24) or
                   (Cardinal(data[offset+1]) << 16) or
                   (Cardinal(data[offset+2]) << 8) or
                   Cardinal(data[offset+3]);
end;

{ Calculate the checksum }
function LZG_CalcChecksum(const data: PByte; size: Cardinal): Cardinal;
var
  a, b: Word;
  i: Cardinal;
begin
  a := 1;
  b := 0;
  for i := 0 to size-1 do
  begin
    a := a + Word(data[i]);
    b := b + a;
  end;
  LZG_CalcChecksum := (Cardinal(b) << 16) or a;
end;

{ Determine the size of the decoded data for a given LZG coded buffer }
function LZG_DecodedSize(const data: PByte; size: Cardinal): Cardinal;
begin
  { Check header }
  if (size < 7) or (data[0] <> Byte('L')) or (data[1] <> Byte('Z')) or
     (data[2] <> Byte('G')) then
  begin
    LZG_DecodedSize := 0;
    Exit;
  end;

  { Get output buffer size }
  LZG_DecodedSize := LZG_GetUINT32(data, 3);
end;

{ Decode LZG coded data }
function LZG_Decode(const indata: PByte; insize: Cardinal; outdata: PByte;
  outsize: Cardinal): Cardinal;
var
  src, inEnd, dst, outEnd, copy: PByte;
  symbol, b, b2, m1, m2, m3, m4, method: Byte;
  i, length, offset, encodedSize, decodedSize, checksum: Cardinal;
begin
  LZG_Decode := 0;

  { Check magic ID }
  if (insize < LZG_HEADER_SIZE) or (indata[0] <> Byte('L')) or
     (indata[1] <> Byte('Z')) or (indata[2] <> Byte('G')) then
    Exit;

  { Get header data }
  decodedSize := LZG_GetUINT32(indata, 3);
  encodedSize := LZG_GetUINT32(indata, 7);
  checksum := LZG_GetUINT32(indata, 11);

  { Check sizes }
  if (outsize < decodedSize) or (encodedSize <> (insize - LZG_HEADER_SIZE)) then
    Exit;

  { Check checksum }
  if LZG_CalcChecksum(@indata[LZG_HEADER_SIZE], encodedSize) <> checksum then
    Exit;

  { Initialize the byte streams }
  src := indata + LZG_HEADER_SIZE;;
  inEnd := indata + insize;
  dst := outdata;
  outEnd := outdata + outsize;

  { Check which method to use }
  method := indata[15];
  if method = LZG_METHOD_LZG1 then
  begin
    { Get marker symbols }
    if not ((src + 4) <= inEnd) then Exit;
    m1 := src[0];
    m2 := src[1];
    m3 := src[2];
    m4 := src[3];
    src := src + 4;

    { Main decompression loop }
    while src < inEnd do
    begin
      symbol := src^;
      Inc(src);

      if (symbol <> m1) and (symbol <> m2) and (symbol <> m3) and (symbol <> m4) then
      begin
        { Literal copy }
        if not (dst < outEnd) then Exit;
        dst^ := symbol;
        Inc(dst);
      end
      else
      begin
        if not (src < inEnd) then Exit;
        b := src^;
        Inc(src);
        if b <> 0 then
        begin
          if symbol = m1 then
          begin
            { marker1 - "Distant copy" }
            if not ((src + 2) <= inEnd) then Exit;
            length := LZG_LENGTH_DECODE_LUT[b and $1f];
            b2 := src^;
            Inc(src);
            offset := ((Cardinal(b and $e0) << 11) or
                       (Cardinal(b2) << 8) or
                       src^) + 2056;
            Inc(src);
          end
          else if symbol = m2 then
          begin
            { marker2 - "Medium copy" }
            if not (src < inEnd) then Exit;
            length := LZG_LENGTH_DECODE_LUT[b and $1f];
            b2 := src^;
            Inc(src);
            offset := ((Cardinal(b and $e0) << 3) or b2) + 8;
          end
          else if symbol = m3 then
          begin
            { marker3 - "Short copy" }
            length := Cardinal(b >> 6) + 3;
            offset := Cardinal(b and $3f) + 8;
          end
          else
          begin
            { marker4 - "Near copy (incl. RLE)" }
            length := LZG_LENGTH_DECODE_LUT[b and $1f];
            offset := Cardinal(b >> 5) + 1;
          end;

          { Copy the corresponding data from the history window }
          copy := dst - offset;
          if not ((copy >= outdata) and ((dst + length) <= outEnd)) then Exit;
          for i := 0 to length-1 do
          begin
            dst^ := copy^;
            Inc(copy);
            Inc(dst);
          end;
        end
        else
        begin
          { Literal copy (single occurance of a marker symbol) }
          if not (dst < outEnd) then Exit;
          dst^ := symbol;
          Inc(dst);
        end;
      end;
    end;
  end
  else if method = LZG_METHOD_COPY then
  begin
    { Plain copy }
    while (src < inEnd) and (dst < outEnd) do
    begin
      dst^ := src^;
      Inc(src);
      Inc(dst);
    end;
  end;

  { All OK? }
  if Cardinal(dst - outdata) = decodedSize then
      LZG_Decode := decodedSize;
end;

end.

